File: general_programming_resources/02_code_organization_modularization/examples/coupled_drivetrain/java/HeadingAdjuster.java
import java.util.ArrayList;
import java.util.List;

public class HeadingAdjuster {
    public static void main(String[] args) {
        Drivetrain dt = new Drivetrain();
        dt.leftMotorPower = 0.8;
        dt.rightMotorPower = 0.8;

        adjustHeading(dt);
        adjustHeading(dt);
        adjustHeading(dt);

        System.out.println("Final left power: " + dt.leftMotorPower);
        System.out.println("Power history: " + dt.powerHistory);
    }

    // Tightly coupled: reaches directly into Drivetrain's public fields
    static void adjustHeading(Drivetrain dt) {
        dt.leftMotorPower = dt.leftMotorPower * 0.9;
        dt.powerHistory.add(dt.leftMotorPower);
    }
}

class Drivetrain {
    public double leftMotorPower;
    public double rightMotorPower;
    public List<Double> powerHistory = new ArrayList<>();
}

File: general_programming_resources/02_code_organization_modularization/examples/coupled_drivetrain/python/heading_adjuster.py
class Drivetrain:
    def __init__(self):
        self.left_motor_power = 0.0
        self.right_motor_power = 0.0
        self.power_history = []


# Tightly coupled: reaches directly into Drivetrain's public attributes
def adjust_heading(dt):
    dt.left_motor_power = dt.left_motor_power * 0.9
    dt.power_history.append(dt.left_motor_power)


def main():
    dt = Drivetrain()
    dt.left_motor_power = 0.8
    dt.right_motor_power = 0.8

    adjust_heading(dt)
    adjust_heading(dt)
    adjust_heading(dt)

    print(f"Final left power: {dt.left_motor_power}")
    print(f"Power history: {dt.power_history}")


if __name__ == "__main__":
    main()
